SignalServiceKit/src/Util/OWSBackgroundTask.m

```diff
@@ -4,14 +4,290 @@
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedMyManager = [[self alloc] initDefault];
+    });
+    return sharedMyManager;
+}
+
+- (instancetype)initDefault
+{
+    OWSAssertIsOnMainThread();
+
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    self.backgroundTaskId = UIBackgroundTaskInvalid;
+    self.expirationMap = [NSMutableDictionary new];
+    self.idCounter = 0;
+    self.isAppActive = CurrentAppContext().isMainAppAndActive;
+
+    OWSSingletonAssert();
+
+    [self observeNotifications];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)observeNotifications
+{
+    if (!CurrentAppContext().isMainApp) {
+        return;
+    }
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationDidBecomeActive:)
+                                                 name:OWSApplicationDidBecomeActiveNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationWillResignActive:)
+                                                 name:OWSApplicationWillResignActiveNotification
+                                               object:nil];
+}
+
+- (void)applicationDidBecomeActive:(UIApplication *)application
+{
+    OWSAssertIsOnMainThread();
+
+    @synchronized(self)
+    {
+        self.isAppActive = YES;
+
+        [self ensureBackgroundTaskState];
+    }
+}
+
+- (void)applicationWillResignActive:(UIApplication *)application
+{
+    OWSAssertIsOnMainThread();
+
+    @synchronized(self)
+    {
+        self.isAppActive = NO;
+
+        [self ensureBackgroundTaskState];
+    }
+}
+
+// Returns nil if adding this task _should have_ started a
+// background task, but the background task couldn't be begun.
+// In that case expirationBlock will not be called.
+- (nullable OWSTaskId)addTask:(BackgroundTaskExpirationBlock)expirationBlock
+{
+    OWSAssert(expirationBlock);
+
+    OWSTaskId _Nullable taskId;
+
+    @synchronized(self)
+    {
+        self.idCounter = self.idCounter + 1;
+        taskId = @(self.idCounter);
+        self.expirationMap[taskId] = expirationBlock;
+
+        if (![self ensureBackgroundTaskState]) {
+            [self.expirationMap removeObjectForKey:taskId];
+            return nil;
+        }
+
+        [self.continuityTimer invalidate];
+        self.continuityTimer = nil;
+
+        return taskId;
+    }
+}
+
+- (void)removeTask:(OWSTaskId)taskId
+{
+    OWSAssert(taskId);
+
+    @synchronized(self)
+    {
+        OWSAssert(self.expirationMap[taskId] != nil);
+
+        [self.expirationMap removeObjectForKey:taskId];
+
+        // Keep the background task active (if necessary) for an
+        // extra fraction of a second to provide continuity between
```

Comment:
Sorry, my comment was confusing.  It now reads:

```
        // This timer will ensure that we keep the background task active (if necessary)
        // for an extra fraction of a second to provide continuity between tasks.
        // This makes it easier and safer to use background tasks, since most code
        // should be able to ensure background tasks by "narrowly" wrapping
        // their core logic with a OWSBackgroundTask and not worrying about "hand off"
        // between OWSBackgroundTasks.
```
So this logic is dealing with that concern, not putting that responsibility on to the shoulders of code that uses `OWSBackgroundTask`.